# Car Price Project

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt

## Data preparation

In [3]:
data = "https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv"

In [4]:
!wget $data -O car-price.csv 

--2023-09-30 23:23:25--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1475504 (1.4M) [text/plain]
Saving to: ‘car-price.csv’

car-price.csv       100%[===================>]   1.41M  37.4KB/s    in 33s     

2023-09-30 23:23:58 (43.4 KB/s) - ‘car-price.csv’ saved [1475504/1475504]



In [63]:
df = pd.read_csv('car-price.csv')
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [64]:
#only these columns used for this homework
used_columns = ['Make',
'Model',
'Year',
'Engine HP',
'Engine Cylinders',
'Transmission Type',
'Vehicle Style',
'highway MPG',
'city mpg',
'MSRP'
]

In [65]:
df = df[used_columns]
df

Make       Model  Year  Engine HP  Engine Cylinders  \
0          BMW  1 Series M  2011      335.0               6.0   
1          BMW    1 Series  2011      300.0               6.0   
2          BMW    1 Series  2011      300.0               6.0   
3          BMW    1 Series  2011      230.0               6.0   
4          BMW    1 Series  2011      230.0               6.0   
...        ...         ...   ...        ...               ...   
11909    Acura         ZDX  2012      300.0               6.0   
11910    Acura         ZDX  2012      300.0               6.0   
11911    Acura         ZDX  2012      300.0               6.0   
11912    Acura         ZDX  2013      300.0               6.0   
11913  Lincoln      Zephyr  2006      221.0               6.0   

      Transmission Type  Vehicle Style  highway MPG  city mpg   MSRP  
0                MANUAL          Coupe           26        19  46135  
1                MANUAL    Convertible           28        19  40650  
2                MANUAL          Coupe           28        20  36350  
3                MANUAL          Coupe           28        18  29450  
4                MANUAL    Convertible           28        18  34500  
...                 ...            ...          ...       ...    ...  
11909         AUTOMATIC  4dr Hatchback           23        16  46120  
11910         AUTOMATIC  4dr Hatchback           23        16  56670  
11911         AUTOMATIC  4dr Hatchback           23        16  50620  
11912         AUTOMATIC  4dr Hatchback           23        16  50920  
11913         AUTOMATIC          Sedan           26        17  28995  

[11914 rows x 10 columns]

In [66]:
#standarize column name
df.columns = df.columns.str.replace(' ', '_').str.lower()
df

make       model  year  engine_hp  engine_cylinders  \
0          BMW  1 Series M  2011      335.0               6.0   
1          BMW    1 Series  2011      300.0               6.0   
2          BMW    1 Series  2011      300.0               6.0   
3          BMW    1 Series  2011      230.0               6.0   
4          BMW    1 Series  2011      230.0               6.0   
...        ...         ...   ...        ...               ...   
11909    Acura         ZDX  2012      300.0               6.0   
11910    Acura         ZDX  2012      300.0               6.0   
11911    Acura         ZDX  2012      300.0               6.0   
11912    Acura         ZDX  2013      300.0               6.0   
11913  Lincoln      Zephyr  2006      221.0               6.0   

      transmission_type  vehicle_style  highway_mpg  city_mpg   msrp  
0                MANUAL          Coupe           26        19  46135  
1                MANUAL    Convertible           28        19  40650  
2                MANUAL          Coupe           28        20  36350  
3                MANUAL          Coupe           28        18  29450  
4                MANUAL    Convertible           28        18  34500  
...                 ...            ...          ...       ...    ...  
11909         AUTOMATIC  4dr Hatchback           23        16  46120  
11910         AUTOMATIC  4dr Hatchback           23        16  56670  
11911         AUTOMATIC  4dr Hatchback           23        16  50620  
11912         AUTOMATIC  4dr Hatchback           23        16  50920  
11913         AUTOMATIC          Sedan           26        17  28995  

[11914 rows x 10 columns]

In [67]:
df.isnull().sum()

make                  0
model                 0
year                  0
engine_hp            69
engine_cylinders     30
transmission_type     0
vehicle_style         0
highway_mpg           0
city_mpg              0
msrp                  0
dtype: int64

In [70]:
#handle missing value with 0
df.engine_hp = df.engine_hp.fillna(0)
df.engine_cylinders = df.engine_cylinders.fillna(0)

df.isnull().sum()

make                 0
model                0
year                 0
engine_hp            0
engine_cylinders     0
transmission_type    0
vehicle_style        0
highway_mpg          0
city_mpg             0
msrp                 0
dtype: int64

In [73]:
#rename "msrp" column to "price"
df.rename(columns={"msrp": "price"}, inplace=True)
df

make       model  year  engine_hp  engine_cylinders  \
0          BMW  1 Series M  2011      335.0               6.0   
1          BMW    1 Series  2011      300.0               6.0   
2          BMW    1 Series  2011      300.0               6.0   
3          BMW    1 Series  2011      230.0               6.0   
4          BMW    1 Series  2011      230.0               6.0   
...        ...         ...   ...        ...               ...   
11909    Acura         ZDX  2012      300.0               6.0   
11910    Acura         ZDX  2012      300.0               6.0   
11911    Acura         ZDX  2012      300.0               6.0   
11912    Acura         ZDX  2013      300.0               6.0   
11913  Lincoln      Zephyr  2006      221.0               6.0   

      transmission_type  vehicle_style  highway_mpg  city_mpg  price  
0                MANUAL          Coupe           26        19  46135  
1                MANUAL    Convertible           28        19  40650  
2                MANUAL          Coupe           28        20  36350  
3                MANUAL          Coupe           28        18  29450  
4                MANUAL    Convertible           28        18  34500  
...                 ...            ...          ...       ...    ...  
11909         AUTOMATIC  4dr Hatchback           23        16  46120  
11910         AUTOMATIC  4dr Hatchback           23        16  56670  
11911         AUTOMATIC  4dr Hatchback           23        16  50620  
11912         AUTOMATIC  4dr Hatchback           23        16  50920  
11913         AUTOMATIC          Sedan           26        17  28995  

[11914 rows x 10 columns]

## Question 1
What is the most frequent observation (mode) for the column transmission_type?


In [79]:
df['transmission_type'].value_counts()

AUTOMATIC           8266
MANUAL              2935
AUTOMATED_MANUAL     626
DIRECT_DRIVE          68
UNKNOWN               19
Name: transmission_type, dtype: int64

The mode from "tranmission_type" column is **AUTOMATIC** with 8266 total data

## Question 2
Create the correlation matrix for the numerical features of your dataset. In a correlation matrix, you compute the correlation coefficient between every pair of features in the dataset.

What are the two features that have the biggest correlation in this dataset?

In [81]:
df.dtypes

make                  object
model                 object
year                   int64
engine_hp            float64
engine_cylinders     float64
transmission_type     object
vehicle_style         object
highway_mpg            int64
city_mpg               int64
price                  int64
dtype: object

In [87]:
numerical_features = ['year','engine_hp','engine_cylinders','highway_mpg','city_mpg']

In [90]:
df[numerical_features].corr()

year  engine_hp  engine_cylinders  highway_mpg  city_mpg
year              1.000000   0.338714         -0.040708     0.258240  0.198171
engine_hp         0.338714   1.000000          0.774851    -0.415707 -0.424918
engine_cylinders -0.040708   0.774851          1.000000    -0.614541 -0.587306
highway_mpg       0.258240  -0.415707         -0.614541     1.000000  0.886829
city_mpg          0.198171  -0.424918         -0.587306     0.886829  1.000000

the biggest  correlation of two features in dataset is between **higway_mpg and city_mpg** with 0.887 corr value